packages/server-wallet/src/models/ledger-proposal.ts

```diff
@@ -0,0 +1,74 @@
+import {JSONSchema, Model, TransactionOrKnex} from 'objection';
+import {Address, SimpleAllocation} from '@statechannels/wallet-core';
+
+import {Bytes32} from '../type-aliases';
+
+interface RequiredColumns {
+  readonly channelId: Bytes32;
+  readonly proposal: SimpleAllocation | null;
+  readonly nonce: number;
+  readonly signingAddress: Address;
+}
+
+const REQUIRED_COLUMNS = ['channelId', 'proposal', 'nonce', 'signingAddress'] as const;
+
+export class LedgerProposal extends Model implements RequiredColumns {
+  static tableName = 'ledger_proposals';
+
+  readonly channelId!: Bytes32;
+  readonly proposal!: SimpleAllocation | null;
+  readonly nonce!: number;
+  readonly signingAddress!: Address;
+
+  static get idColumn(): string {
+    return 'channelId';
+  }
+
+  static get jsonSchema(): JSONSchema {
+    return {
+      type: 'object',
+      required: [...REQUIRED_COLUMNS],
+    };
+  }
+
+  static async forChannel(
+    channelId: Bytes32,
+    txOrKnex: TransactionOrKnex
+  ): Promise<LedgerProposal[]> {
+    return await LedgerProposal.query(txOrKnex).where({channelId});
+  }
+
+  static async setProposalsToNull(channelId: Bytes32, txOrKnex: TransactionOrKnex): Promise<void> {
+    await LedgerProposal.query(txOrKnex)
+      .where({channelId})
+      .patch({proposal: null});
+  }
+
+  static async storeProposal(
+    {channelId, signingAddress, proposal, nonce}: RequiredColumns,
+    txOrKnex: TransactionOrKnex
+  ): Promise<void> {
+    // TODO: Do this in a single SQL query (could not figure out how in time)
```

Comment:
Could use help if anyone knows how. Tried 2 strategies:

Strategy 1 — raw SQL
```ts
await (tx || this.knex).raw(`
  INSERT INTO ${LedgerProposal.tableName}
    (channel_id, signing_address, proposal, nonce)
      values ("${channelId}", "${signingAddress}", ${JSON.stringify(proposal)}, ${nonce}
  ON CONFLICT (channel_id, signing_address)       //   ^-- unsure how to do jsonb insertion
  DO
    UPDATE
    WHERE proposal IS NOT NULL
    AND nonce > ${nonce}
    SET proposal = "${JSON.stringify(proposal)}", nonce = ${nonce};`);
```

Strategy 2 — use .onConflict() API
```ts

await (tx || this.knex)(LedgerProposal.tableName)
  .insert({channelId, signingAddress, proposal, nonce})
  .onConflict(['channelId', 'signingAddress'])
  .merge({proposal, nonce})
  .whereNull('proposal')           //   <-- I don't think this applies in the
  .andWhere('nonce', '<', nonce);  //          right place for the query
```

Reply:
Are you saying that
1. fails because you are not inserting a jsonb column correctly
2. fails because you think the `whereNull` is not in the correct place
?

Reply:
(1) I wasn't sure how to write the query to insert an `object` and coerce it into a `jsonb` literal inside the `INSERT` and `SET` command, so I tried using the Knex API but

(2) It didn't seem like the ORM created the right SQL query to do what I want (which is what the raw SQL query was doing), so two concurrency related tests failed

—

I simply stopped trying and wrote two queries with the nonce check done in JS. Have not revisited this since last night

Reply:
The issue goes away with the "soft-delete" strategy of having a `dismissed: boolean` column, since then you simply insert a new row for new proposals.